### PR TITLE
add a hashchain example (Nova Forward port)

### PR DESCRIPTION
**Pull request by huitseeker**
This forward ports the following PR:
https://github.com/microsoft/Nova/pull/311

### CODE REVIEW

**Comment by huitseeker** [comment]
!benchmark --bench recursive-snark --bench recursive-snark-supernova --features cuda,asm

**Comment by huitseeker** [comment]
!benchmark --bench recursive-snark --bench recursive-snark-supernova

**Comment by samuelburnham** [comment]
!benchmark --bench recursive-snark --bench recursive-snark-supernova --features cuda,asm

File: examples/hashchain.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+      let output = neptune::sponge::api::SpongeAPI::squeeze(&mut sponge, 1, acc);
+      sponge.finish(acc).unwrap();
+      Elt::ensure_allocated(&output[0], &mut ns.namespace(|| "ensure allocated"), true)?
+    };
+
+    Ok(vec![z_out])
+  }
+}
+
+/// cargo run --release --example and
```
Comment (tchataigner) [inline examples/hashchain.rs:106]:
```suggestion
/// cargo run --release --example hashchain
```

**Review by tchataigner — APPROVED** [review]
LGTM! There is only one typo in the comment to run the example

**Review by tchataigner — APPROVED** [review]
LGTM

**Pull request merged**